Updates fixes

huntrese opened the pull request:
Modied ifs into early return,
Used Relationships where possible,
Switched to using Auth middleware for validation,
added migration from sql dump, not 100% sure it works

huntrese committed:
Switch to Auth Middleware, Use Relationships

huntrese committed:
add migrations

flaviu-chelaru commented on app/Http/Controllers/Auth/RegisterController.php:
```diff
@@ -68,6 +68,8 @@ protected function create(array $data)
             'name' => $data['name'],
             'email' => $data['email'],
             'password' => Hash::make($data['password']),
+            'avatar' => "https://www.seiu1000.org/sites/main/files/main-images/camera_lense_0.jpeg",
```
https://picsum.photos/ - use a lorem picture service. they are public.

flaviu-chelaru commented on app/Http/Controllers/AuthorController.php:
```diff
@@ -0,0 +1,22 @@
+    
+        $books =$author->books;
+    
+        return view('Author', compact('author', 'books'));
```
some operating systems will be case sensitive. be careful to this

flaviu-chelaru commented on app/Http/Controllers/AvatarController.php:
```diff
@@ -11,48 +11,28 @@ class AvatarController extends Controller
+        $data = $request->all();
+        $account->name = $data['name'];
+        $account->email = $data['email'];
+        $account->avatar = $data['avatar'];
```
this might be a file upload. take it into consideration

General:
if you are working with arrays, you must always assume that the key might not exist.

instead of $data['name'], use $data['name'] ?? ''

Specific

Specific to you case here, don't use $data = $request->all().

Instead use $account->name = $request->post->get('name') ( i believe this was the correct method call)

flaviu-chelaru commented on app/Http/Controllers/AvatarController.php:
```diff
@@ -11,48 +11,28 @@ class AvatarController extends Controller
 
+       
+    }
+    // app/Http/Controllers/AccountController.php
```
remove the comment :D

flaviu-chelaru commented on app/Http/Controllers/ChapterController.php:
```diff
@@ -12,65 +12,72 @@
-                ->update(['chapter' => $chapter_no]);
+        // Update recent entry if user is logged in
+        if ($user_id) {
+            $recent = $user->recent()->where("user_id", $user_id)->first();
```
lines 20 - 33 - these belong to a subscriber / events. i would say to remove it for now, and after a chapter is selected successfully just fire an event(new ChapterOpened(User $user, Chapter $chapter))

We will write a subscriber - you will see it has a much cleaner code :)

flaviu-chelaru commented on app/Http/Controllers/ChapterController.php:
```diff
@@ -12,65 +12,72 @@
-
+        $book = Books::find($book_id);
     
+        if (!$book) {
```
use Model::findOrFail instead - it has the same purpose

flaviu-chelaru commented on app/Http/Controllers/AuthorController.php:
```diff
@@ -0,0 +1,22 @@
+    {
+        $author = Authors::find($author_id);
+    
+        if (!$author) {
```
use Model::findOrFail instead - it has the same purpose

flaviu-chelaru commented on app/Http/Controllers/ChapterController.php:
```diff
@@ -12,65 +12,72 @@
     
+        // Calculate previous and next chapter numbers
+        $prev = ($chapter_no != 1) ? $chapter_no - 1 : -1;
+        $next = ($chapters->count() > $chapter_no) ? $chapter_no + 1 : -1;
```
you can use the laravel standard for paginations - https://laravel.com/docs/10.x/pagination#converting-results-to-json

flaviu-chelaru commented on app/Http/Controllers/ChapterController.php:
```diff
@@ -12,65 +12,72 @@
+    
+    // Logic to determine the URL for reading
+    $url = "books/$book_id/$chapter";
+    return redirect($url);
```
please use named redirects - https://laravel.com/docs/10.x/redirects#redirecting-named-routes

flaviu-chelaru commented on app/Http/Controllers/LibraryController.php:
```diff
@@ -9,29 +9,38 @@
+    {
+        
+        $book_id = $request->input('book_ID');
+        // dd($book_ID);
```
remove this kind of comments, please :D

flaviu-chelaru commented on app/Http/Controllers/LibraryController.php:
```diff
@@ -9,29 +9,38 @@
+        $library = $user->library;
+        $bookIDs = $library->pluck('book_ID')->toArray();
+        $isInLibrary = in_array($book_id, $bookIDs);   
+        if ($isInLibrary) {
```
if it is already in the library, add it again? If it is not, then don't add it? 

I believe this is not the intented behaviour

flaviu-chelaru commented on app/Http/Controllers/RecentController.php:
```diff
@@ -12,37 +12,30 @@ class RecentController extends Controller
         $user = Auth::user();
-        if(Auth::check()){
-            $recentBooks = Recent::where('user_id', Auth::user()->user_ID)->orderBy('updated_at', 'desc')->get();
+        $recentBooks = $user->recent()->orderBy('updated_at', 'desc')->get();
```
you can achieve similar result if you use the fluent relatinship select 

The foreach below - this is the n+1 problem in sql - a very common and dangerous pitfall of performance

https://planetscale.com/blog/what-is-n-1-query-problem-and-how-to-solve-it

flaviu-chelaru commented on app/Http/Controllers/SearchController.php:
```diff
@@ -15,10 +15,8 @@ public function search(Request $request)
-                    ->get();
-        $author = Authors::where('name', 'LIKE', '%' . $q . '%')
-                    ->get();
+        $book = Books::where('name', 'LIKE', '%' . $q . '%')->get();
```
try something like this

$book = Books::where()->select([col1, col2, DB::raw('"book" as type')]);
$author = Author::where()->select(/* identical as above, with different type*/)->union($book);

$autor->paginate();

Should give you a nice result where you no longer need to do if statements and count on objects (array of objects)

Small suggestion

flaviu-chelaru commented on app/Http/Middleware/Authenticate.php:
```diff
@@ -12,6 +12,6 @@ class Authenticate extends Middleware
     protected function redirectTo(Request $request): ?string
     {
-        return $request->expectsJson() ? null : route('login');
+        return $request->expectsJson() ? null : route('please_login');
```
i like the addition of "please" =)) Just make sure the name of the route is correct :D

flaviu-chelaru commented on app/Models/Users.php:
```diff
@@ -11,9 +13,20 @@ class Users extends Authenticatable
+    {
+        return $this->hasMany(Library::class,"user_ID","user_ID");
+    }
+    public function booksInLibrary()
```
advice: stay consistent. If you defined return types to other methods, define it for all

flaviu-chelaru commented on database/migrations/2023_11_01_162501_create_db.php:
```diff
@@ -11,13 +11,9 @@
-            $table->timestamps(); // Add this line if you want to enable timestamps
-        });
+       DB::unprepared(
+        file_get_contents(__DIR__."")
```
this is a bit dangerous :D

flaviu-chelaru replied:
https://laravel-news.com/migration-generator-for-laravel 

Use that tool

- Require-dev the tool
- delete all existing migrations
- run the tool
- remove the tool after the migrations have been created.

flaviu-chelaru commented on resources/views/author.blade.php:
```diff
@@ -0,0 +1,69 @@
+<style>
+     .book-card {
```
don't inline your css.. keep it externally. 
If you don't want to use bootstrap css as a library, pico css or tailwind can give you a very basic css for your project. 

Take a look https://picocss.com/

flaviu-chelaru commented on resources/views/author.blade.php:
```diff
@@ -0,0 +1,69 @@
+            @if(isset($books))
+            @foreach($books as $book)
+            <div class="book-card">
+                <a href="/books/{{$book->book_ID}}" class="book-link">
```
use the blade way for generating the urls. It's safer this way

huntrese committed:
Adressing Issues from PR

huntrese committed:
create migrations,
swap early returns for findOrFail,
update userProfile post,
use joins instead of multiple querries
externalized the css

huntrese commented:
updates:

swap early returns for findOrFail,
update userProfile post,
use joins instead of multiple querries
externalized the css

not yet done:
pagination, events

huntrese committed:
Merge branch 'updates_Fixes' of https://github.com/huntrese/Ebooks into updates_Fixes

huntrese commented:
tried to rename last commit to something more sensible... failed terribly
)

flaviu-chelaru commented on composer.json:
```diff
@@ -15,6 +15,7 @@
         "phpmyadmin/phpmyadmin": "^4.9"
     },
     "require-dev": {
+        "bennett-treptow/laravel-migration-generator": "^4.3",
```
asta nu il pastram. asta era sa il folosesti cat generezi migrari, apoi il scoti afara.. asta ii de executat o singura data, dupa aia ar fi de preferat sa te obisnuiesti tu sa scrii fisiere de migrare in laravel.

flaviu-chelaru commented on resources/views/author.blade.php:
```diff
@@ -50,12 +16,13 @@
             @foreach($books as $book)
             <div class="book-card">
-                <a href="/books/{{$book->book_ID}}" class="book-link">
+                <a href="{{url('books/' . $book->book_ID)}}" class="book-link">
```
foloste named routes, te rog. numele route-ului se schimba mai rar decat route-ul in sine (de obicei)

huntrese committed:
remove unnecessary requirements,
use name route isntead of url
